pkg/mcclient/mcclient.go

```diff
@@ -21,46 +21,47 @@ package mcclient
 import (
 	"context"
 	"errors"
-	"reflect"
+	"fmt"
 	"sync"
+	"time"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/fields"
-	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/tools/clientcmd"
-	log "k8s.io/klog/v2"
 
 	"github.com/kcp-dev/edge-mc/pkg/apis/edge/v1alpha1"
 	ksclientset "github.com/kcp-dev/edge-mc/pkg/client/clientset/versioned"
 	ksinformers "github.com/kcp-dev/edge-mc/pkg/client/informers/externalversions"
 	mcclientset "github.com/kcp-dev/edge-mc/pkg/mcclient/clientset"
-	"github.com/kcp-dev/edge-mc/pkg/mcclient/listwatch"
 )
 
 type KubestellarClusterInterface interface {
 	// Cluster returns clientset for given cluster.
 	Cluster(name string) (client mcclientset.Interface)
 	// ConfigForCluster returns rest config for given cluster.
 	ConfigForCluster(name string) (*rest.Config, error)
-	// CrossClusterListWatch returns cross-cluster ListWatch
-	CrossClusterListWatch(gv schema.GroupVersion, resource string, namespace string, fieldSelector fields.Selector) *listwatch.CrossClusterListerWatcher
 }
 
 type multiClusterClient struct {
-	configs    map[string]*rest.Config
-	clientsets map[string]*mcclientset.Clientset
-	lock       sync.Mutex
+	ctx              context.Context
+	configs          map[string]*rest.Config
+	clientsets       map[string]*mcclientset.Clientset
+	managerClientset *ksclientset.Clientset
+	lock             sync.Mutex
 }
 
 func (mcc *multiClusterClient) Cluster(name string) mcclientset.Interface {
 	mcc.lock.Lock()
 	defer mcc.lock.Unlock()
 	clientset, ok := mcc.clientsets[name]
 	if !ok {
-		//TODO change to ClusterOrDie and panic? return an error?
-		panic("invalid cluster name")
+		// try to get LogicalCluster from API server
+		clientset, err := mcc.getFromServer(name)
```

Comment:
Why don't we cache the config & cs when doing getFromServer ?  If we missed somehow the event you will come here every time someone use this cluster, right ?

Reply:
We do cache it in `getFromServer ` function if it exists and valid.

Reply:
right, the gitdiff confused me with  the ondelete change :(